File: src/cli/context.ts
import { loadConfig, type GateConfig } from "../core/config.js";
import { NO_GIT_BRANCH_KEY, readCurrentRunId, resolveBranchKey } from "../core/current.js";
import { findGateRoot } from "../core/paths.js";
import { readRun, type Run } from "../core/run.js";
import type { ParsedArgs } from "./args.js";
import { GateError } from "./output.js";

export function requireRoot(): string {
  const root = findGateRoot();
  if (!root) throw new GateError("no .gate/ found — run `gate init` first");
  return root;
}

export interface ActiveContext {
  root: string;
  run: Run;
  config: GateConfig;
}

/**
 * Resolve the run a phase-scoped command (check/next/review/retro/skip/log/
 * playbook) operates on: `--run <id>` always wins (works even in detached
 * HEAD - its stated use case); otherwise the active run for the current
 * branch (Milestone 4). A detached HEAD has no branch to resolve "current"
 * from, so it refuses with a pointer to `--run` rather than guessing which
 * run the caller meant.
 *
 * `--run` is not a bare bypass: gates measure `changedFiles(root,
 * run.baseRef)` against whatever is actually checked out, so pointing it at
 * a run whose tree doesn't match the working directory lets TEST/REVIEW pass
 * on a vacuous empty diff (100% coverage, trivially in-scope) while IMPLEMENT
 * alone is protected by its non-empty-diff check. Refuse a non-active run
 * (done/abandoned) outright, and - when the run recorded a branch - refuse a
 * checked-out branch that doesn't match it; a detached HEAD skips that branch
 * check entirely, since there's no checked-out branch to compare against.
 */
export function requireActiveRun(args?: ParsedArgs): ActiveContext {
  const root = requireRoot();
  const config = loadConfig(root);

  const explicit = args && typeof args.flags.run === "string" ? args.flags.run : undefined;
  if (explicit) {
    const run = readRun(root, explicit);
    if (run.status !== "active") {
      throw new GateError(`run "${explicit}" is ${run.status}, not active — --run only selects an in-flight run`);
    }
    const resolved = resolveBranchKey(root);
    if (run.branch && resolved.kind === "key" && resolved.key !== run.branch) {
      throw new GateError(
        `run "${explicit}" was started on branch "${run.branch}", but "${resolved.key}" is checked out — ` +
          `checkout "${run.branch}" first (or detach HEAD) to act on this run`,
      );
    }
    return { root, run, config };
  }

  const resolved = resolveBranchKey(root);
  if (resolved.kind === "detached") {
    throw new GateError(
      "HEAD is detached — no branch to resolve the active run from; pass --run <id> (see `gate status` for runs in flight)",
    );
  }
  const id = readCurrentRunId(root, resolved.key);
  if (!id) {
    throw new GateError(
      resolved.key === NO_GIT_BRANCH_KEY
        ? "no active run — start one with `gate start \"<title>\"`"
        : `no active run on branch "${resolved.key}" — start one with \`gate start "<title>"\``,
    );
  }
  return { root, run: readRun(root, id), config };
}

File: rust/src/cli/context.rs
//! Port of `src/cli/context.ts`: resolving the `.gate/` root and the run a
//! phase-scoped command operates on.

use std::env;
use std::path::PathBuf;

use crate::cli::args::ParsedArgs;
use crate::cli::output::UserError;
use crate::core::config::{load_config, GateConfig};
use crate::core::current::{
    read_current_run_id, resolve_branch_key, BranchKeyResolution, NO_GIT_BRANCH_KEY,
};
use crate::core::paths::{find_gate_root, validate_run_id};
use crate::core::run::{read_run, Run, RunStatus};

/// Port of `requireRoot`.
pub fn require_root() -> Result<PathBuf, UserError> {
    let cwd = env::current_dir().map_err(|e| UserError::new(e.to_string()))?;
    find_gate_root(&cwd).ok_or_else(|| UserError::new("no .gate/ found - run `gate init` first"))
}

/// Port of `ActiveContext`.
pub struct ActiveContext {
    pub root: PathBuf,
    pub run: Run,
    pub config: GateConfig,
}

/// Port of `requireActiveRun`: `--run <id>` always wins (works even in
/// detached HEAD - its stated use case); otherwise the active run for the
/// current branch. A detached HEAD has no branch to resolve "current" from,
/// so it refuses with a pointer to `--run` rather than guessing. A `--run`
/// pointing at a non-active run, or at a run recorded on a branch other than
/// the checked-out one, is refused (see the TS doc comment for why `--run`
/// is not a bare bypass).
pub fn require_active_run(args: Option<&ParsedArgs>) -> Result<ActiveContext, UserError> {
    let root = require_root()?;
    let config = load_config(&root)?;

    let explicit = args.and_then(|a| a.flags.str("run")).map(str::to_string);
    if let Some(explicit) = explicit {
        validate_run_id(&explicit)?;
        let run = read_run(&root, &explicit)?;
        if run.status != RunStatus::Active {
            return Err(UserError::new(format!(
                "run \"{}\" is {}, not active - --run only selects an in-flight run",
                explicit,
                run.status.as_str()
            )));
        }
        let resolved = resolve_branch_key(&root);
        if let (Some(branch), BranchKeyResolution::Key(key)) = (&run.branch, &resolved) {
            if key != branch {
                return Err(UserError::new(format!(
                    "run \"{explicit}\" was started on branch \"{branch}\", but \"{key}\" is checked out - checkout \"{branch}\" first (or detach HEAD) to act on this run"
                )));
            }
        }
        return Ok(ActiveContext { root, run, config });
    }

    let resolved = resolve_branch_key(&root);
    let key = match resolved {
        BranchKeyResolution::Detached => {
            return Err(UserError::new(
                "HEAD is detached - no branch to resolve the active run from; pass --run <id> (see `gate status` for runs in flight)",
            ));
        }
        BranchKeyResolution::Key(key) => key,
    };
    let id = read_current_run_id(&root, &key)?;
    let Some(id) = id else {
        return Err(UserError::new(if key == NO_GIT_BRANCH_KEY {
            "no active run - start one with `gate start \"<title>\"`".to_string()
        } else {
            format!("no active run on branch \"{key}\" - start one with `gate start \"<title>\"`")
        }));
    };
    // Gate report finding 4: `id` came from `.gate/current.json`, not free
    // text a caller already validated - reject an absolute/`..` id before
    // it reaches `read_run`/`run_paths`.
    validate_run_id(&id)?;
    let run = read_run(&root, &id)?;
    Ok(ActiveContext { root, run, config })
}
